raiden/network/rpc/client.py

```diff
@@ -563,12 +566,16 @@ def new_filter(
             to_block: typing.BlockSpecification = 'latest',
     ) -> Filter:
         """ Create a filter in the ethereum node. """
-        return self.web3.eth.filter({
-            'fromBlock': from_block,
-            'toBlock': to_block,
-            'address': to_normalized_address(contract_address),
-            'topics': topics,
-        })
+        return StatelessFilter(
+            self.web3,
+            {
+                'fromBlock': from_block,
+                'toBlock': to_block,
+                'address': to_normalized_address(contract_address),
+                'topics': topics,
+            },
+            get_block_number=lambda: self.last_block_number,
```

Comment:
This is a really bad idea IMO, it's adding a hidden contract, were for the filters to work properly the rpc client `block_number` method must be used, otherwise the cache is not updated and new events will not be polled.

For this to work properly this is required:

1. Because the cache is only updated on calls to `JSONRPCClient.block_number`, this method must be called
2. For the above, we rely on [`BlockChainService.block_number()`](https://github.com/raiden-network/raiden/blob/6444580a4208a32afcebc5415c4c34db671967c2/raiden/network/blockchain_service.py#L49-L50) to always forward the call to it's client instance.
3. For the above to happen, we rely on the [alarm task to call chain.block_number()](https://github.com/raiden-network/raiden/blob/6444580a4208a32afcebc5415c4c34db671967c2/raiden/tasks.py#L79)

If we change 2 or 3 (e.g. replace by a direct call to the web3) the cache will not be updated and the filter will stop working, if this happens it will be a hard to debug problem.

Reply:
@hackaugusto yes, I agree this contract exists in the current implementation, but I don't see a better way to have an up-to-date `block_number` in the filter: if we want to avoid the double-call, we need to rely on **someone** knowing the latest block number.
This was the best approach IMO because the client is both the one who gets to do all the `blockNumber` calls (currently) and also the one who instantiates the filter objects.
As a simple `latest` query doesn't atomically inform **which** block was latest at the moment of the call, an alternative would be to use the last event blockNumber (if any), but it's sub-optimal.

Reply:
> but I don't see a better way to have an up-to-date block_number in the filter: if we want to avoid the double-call, we need to rely on someone knowing the latest block number.

That is why I'm proposing that the block number should be passed as an argument, because caller is the one that can know what is the current view of the block number.